app/services/licences/fetch-licence-history.service.js

```diff
@@ -0,0 +1,89 @@
+'use strict'
+
+/**
+ * Fetches data needed for the view '/licences/{id}/history` page
+ * @module FetchLicenceHistoryService
+ */
+
+const { db } = require('../../../db/db.js')
+
+const ChargeVersionModel = require('../../models/charge-version.model.js')
+const LicenceModel = require('../../models/licence.model.js')
+const LicenceVersionModel = require('../../models/licence-version.model.js')
+const ReturnVersionModel = require('../../models/return-version.model.js')
+
+/**
+ * Fetches data needed for the view '/licences/{id}/history` page
+ *
+ * @param {string} licenceId - The UUID for the licence to fetch
+ *
+ * @returns {Promise<module:LicenceModel|ChargeVersionModel|LicenceVersionModel|ReturnVersionModel>} the licence and
+ * related charge, licence and return versions
+ */
+async function go (licenceId) {
+  const licence = await _fetchLicence(licenceId)
+  const chargeVersions = await _fetchChargeVersions(licenceId)
+  const licenceVersions = await _fetchLicenceVersions(licenceId)
+  const returnVersions = await _fetchReturnVersions(licenceId)
+
+  return {
+    entries: [...chargeVersions, ...licenceVersions, ...returnVersions],
+    licence
+  }
+}
+
+async function _fetchLicence (licenceId) {
+  return LicenceModel.query()
+    .findById(licenceId)
+    .select([
+      'id',
+      'licenceRef',
+      'createdAt'
+    ])
+}
+
+async function _fetchChargeVersions (licenceId) {
+  return ChargeVersionModel.query()
+    .select(
+      db.raw("'charge-version' as entry_type"),
+      'chargeVersions.id as entry_id',
+      'chargeVersions.createdAt as created_at'
+    )
+    .modify('history')
+    .orderBy([
+      { column: 'versionNumber', order: 'desc' }
+    ])
+    .where('chargeVersions.licenceId', licenceId)
+}
+
+async function _fetchLicenceVersions (licenceId) {
+  return LicenceVersionModel.query()
+    .select(
+      db.raw("'licence-version' as entry_type"),
+      'licenceVersions.id as entry_id',
+      'licenceVersions.createdAt as created_at'
+    )
+    .modify('history')
+    .orderBy([
+      { column: 'issue', order: 'desc' }
+    ])
+    .where('licenceVersions.licenceId', licenceId)
+}
+
+async function _fetchReturnVersions (licenceId) {
+  return ReturnVersionModel.query()
+    .select(
+      db.raw("'return-version' as entry_type"),
+      'returnVersions.id as entry_id',
+      'returnVersions.createdAt as created_at'
+    )
+    .modify('history')
+    .orderBy([
+      { column: 'version', order: 'desc' }
+    ])
+    .where('returnVersions.licenceId', licenceId)
+}
```

Comment:
A few things in this.

## Renaming fields with snake case

Renaming fields, for example, `createdAt` to `created_at` is doing nothing. You can see this in your presenter where you are still referencing everything as `createdAt`. 

We reference a view for all existing tables we use rather than the table directly. Doing so allows us to

- ignore fields that are never used or always populated with the same value
- rename fields to be consistent, for example, `date_created` to `created_at`

All **Objection.js** and **Knex** queries (`db('licences).withSchema('public).select(*)` for example) then have Knex snake case mappers applied (see `knexfile.application.js`). This is what converts the PostgreSQL snake_case field names into JavaScript friendly camelCase.

My source query bypassed the snake case mappers by using `db.raw()` fully (not as a partial within an existing query). Hence, fields would have come back as snake_case.

## Renaming to be consistent across queries

The fields were renamed in my source query to use generic names because the query used a [UNION ALL](https://www.w3schools.com/sql/sql_ref_union_all.asp) to bring the results from across 3 tables into one.

Now the service is firing these queries individually via the Objection.js models, there is no point in doing this.

## Only one query needed

The service is currently fetching the licence, then fetching the charge, licence and return versions in separate queries. But as these are linked to the `LicenceModel` via the relationships we have defined in the model, they could all be obtained with a single Objection.js query to `LicenceModel` and then `withGraphFetched()` for each of the version types.

The service could be fetching a single licence with all its associated versions and their mod logs attached as a single instance to the presenter to them be formatted ready for the view.